io/src/tim_grabber.cpp

```diff
@@ -0,0 +1,265 @@
+  {
+    cos_dynamic_lookup_table_.push_back(std::cos (angle));
+    sin_dynamic_lookup_table_.push_back(std::sin (angle));
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::updateLookupTables () {
+  //// note: cos, sin are the same size
+  if (cos_dynamic_lookup_table_.size () != amount_of_data_)
+    buildLookupTables ();
+}
+
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+bool
+pcl::TimGrabber::isValidPacket () const {
+  return received_packet_.data ()[length_-1] == '\03';
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+void
+pcl::TimGrabber::receiveTimPacket ()
+{
+  if (!tim_socket_.is_open ())
+    return;
+
+  tim_socket_.send (boost::asio::buffer ("\x02sRN LMDscandata 1\x03\0"));
+
+  std::this_thread::sleep_for (std::chrono::milliseconds (wait_time_milliseconds_));
+
+  length_ = tim_socket_.receive (boost::asio::buffer (received_packet_));
+
+  if (!isValidPacket ()) {
+    //// Countup wait_time
+    //// Because fit wait_time to network latency
+    wait_time_milliseconds_++;
+
+    //// Clear invalid packet in socket
+    while (!isValidPacket ())
+      length_ = tim_socket_.receive (boost::asio::buffer (received_packet_));
+
+    //// If received packet is invalid, recurse
+    receiveTimPacket ();
+  }
+}
+
+void
+pcl::TimGrabber::parsePacketHeader (std::istringstream& ss) {
+  //// note: Protcol named CoLaA (used by SICK) has some information.
+  ////       In this Grabber, only the amount_of_data is used, so other information is truncated.
+  ////       Details of the protocol can be found at the following URL.
+
+  //// pp.87~89 (table)
+
+  //// https://cdn.sickcn.com/media/docs/7/27/927/technical_information_telegram_listing_ranging_sensors_lms1xx_lms5xx_tim2xx_tim5xx_tim7xx_lms1000_mrs1000_mrs6000_nav310_ld_oem15xx_ld_lrs36xx_lms4000_en_im0045927.pdf
+
+
+  //// By this PDF, the header contains the following information in order
+
+  /////////////////////////////////////////////////////
+  //// command type
+  //// command
+  //// version number
+  //// device number
+  //// serial number (2 value)
+  //// device status
+  //// Telegram counter
+  //// Scan counter
+  //// Time since start up
+  //// Time of transmission
+  //// Status of digital inputs (2 value)
+  //// Status of digital outputs (2 value)
+  //// Reserved
+  //// scan frequency
+  //// measurement frequency
+  //// Amount of encoder
+  //// Amount of 16 bit channels
+  //// Content
+  //// Scale factor according to IEEE754
+  //// Scale factor offset according to IEEE754
+  //// Start angle
+  //// Size of single angular step
+  //// Amount of data
+  /////////////////////////////////////////////////////
+
+  for (short i=0; i<25; ++i) {
+    std::string str;
+    ss >> str;
+  }
+
+  ss >> std::hex >> amount_of_data_;
```

Comment:
Reading this isn't needed. You can literally reach in to grab the data and save time (see one of my previous message regarding this)

Reply:
Year!

I changed function argument type and applied it.
That reduced the number of inputs to the stream.

In addition, it seemed possible, so I rewrote `setf()`.